Fix objPoints order in GridBoard and CharucoBoard

AleksandrPanov opened the pull request:
Fixes https://github.com/opencv/opencv_contrib/issues/2623 and https://github.com/opencv/opencv_contrib/issues/2604

This PR:

**Fixed objPoints order in GridBoard and CharucoBoard**.
In the `_drawPlanarBoardImpl()` function the y coordiantes of the charuco markers get horizontally mirrored by `pf.y = (1.0f - pf.y / sizeY) * float(out.rows);` Now `pf.y = pf.y / sizeY * float(out.rows);`

Drawing correctness was checked by [example.cpp.txt](https://github.com/opencv/opencv_contrib/files/8074445/example.cpp.txt)

**Before fix:**
- Arrays `objPoints` and `ids` for 3x4 `aruco::GridBoard` are in the wrong order
![board_detected_grid](https://user-images.githubusercontent.com/22337800/154144048-0608b2b7-ed13-4465-84a7-f8adb67b3623.jpg)
- Arrays `objPoints` and `ids` for 3x4 `aruco::CharucoBoard` are in the wrong order. Also `objPoints` coordinates are incorrect.
![charuco_detected_grid](https://user-images.githubusercontent.com/22337800/154145627-b04ddb47-8353-4376-b30c-fb731b798092.jpg)

**After fix:**
![image](https://user-images.githubusercontent.com/22337800/154146231-c9dc9c8b-4602-48bf-bccc-1e2313fe7dbc.png)
Everything is drawn correctly.


**Fixed generation of rotated images (in ArUco tests). Now rotated images are generated using the "[aircraft principal axes](https://en.wikipedia.org/wiki/Aircraft_principal_axes)".**

![image](https://user-images.githubusercontent.com/22337800/153911933-d7857e67-21e1-4d98-97c3-37824c7ceb82.png)
![rotate_pitch](https://user-images.githubusercontent.com/22337800/153911969-a1e64441-6346-425e-90c6-ddbd5def03b9.jpg)
![rotate_yaw](https://user-images.githubusercontent.com/22337800/153911974-79dbf02c-cfdd-4036-9cb7-44af07094201.jpg)

**Also fixed `objPoints[markerIndex]` order (from CCW to CW) in GridBoard and CharucoBoard.** Because tutorial and all ArUco algs use CW order.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented on modules/aruco/include/opencv2/aruco.hpp:
```diff
@@ -303,6 +303,8 @@ class CV_EXPORTS_W Board {
     /// The identifiers refers to the board dictionary
     CV_PROP_RW std::vector< int > ids;
+
+    CV_PROP Point3f rightBottomBorder;
```
Please add documentation line for the property.

AleksandrPanov replied:
added

AleksandrPanov replied:
![image](https://user-images.githubusercontent.com/22337800/154146907-133bbb0a-78eb-4eb4-9d0c-f26d4625bad4.png)

asmorkalov commented on modules/aruco/test/test_aruco_utils.hpp:
```diff
@@ -0,0 +1,102 @@
+
+            // rotate camera in pitch axis
+            Mat rotPitch(3, 1, CV_64FC1);
+            rotPitch.ptr< double >(0)[0] = pitch;
```
rotPitch.at< double >(i, j) is better.

AleksandrPanov replied:
fixed

AleksandrPanov commented:
![rotate_yaw](https://user-images.githubusercontent.com/22337800/153900814-b005d989-5857-4a0c-ae7c-142927c49d89.jpg)

alalek commented on modules/aruco/include/opencv2/aruco.hpp:
```diff
@@ -294,7 +294,7 @@ class CV_EXPORTS_W Board {
 
     /// array of object points of all the marker corners in the board
-    /// each marker include its 4 corners in CCW order. For M markers, the size is Mx4.
+    /// each marker include its 4 corners in CW order. For M markers, the size is Mx4.
```
It doesn't make sense to mention CW or CCW without specifying used coordinate system.

Please take a look on calib3d module documentation.

AleksandrPanov replied:
Added this definition:
```
    /// each marker include its 4 corners in this order:
    ///-   objPoints[i][0] - left-top point of i-th marker
    ///-   objPoints[i][1] - right-top point of i-th marker
    ///-   objPoints[i][2] - right-bottom point of i-th marker
    ///-   objPoints[i][3] - left-bottom point of i-th marker
```

alalek commented on modules/aruco/test/test_aruco_utils.hpp:
```diff
@@ -0,0 +1,101 @@
+#include "test_precomp.hpp"
```
Missing license header

AleksandrPanov replied:
fixed

alalek commented on modules/aruco/src/aruco.cpp:
```diff
@@ -1713,20 +1719,19 @@ Ptr<GridBoard> GridBoard::create(int markersX, int markersY, float markerLength,
-            corners[2] = corners[0] + Point3f(markerLength, -markerLength, 0);
-            corners[3] = corners[0] + Point3f(0, -markerLength, 0);
+            corners[2] = corners[0] + Point3f(markerLength, markerLength, 0);
+            corners[3] = corners[0] + Point3f(0, markerLength, 0);
```
You just changed the coordinate system by flipping Y.

AleksandrPanov replied:
This changes was added to make coordinate systems consistent.
Now in `Board::create()` use this coordinate system:
![image](https://user-images.githubusercontent.com/22337800/154340852-5fdcfc6e-896b-4c7f-a5bc-3559f4d78e59.png)
But `Board::draw()` use another coordinate system:
![image](https://user-images.githubusercontent.com/22337800/154341321-d8f4908e-dffe-45d5-b80e-ac7cbb5da057.png)

Also, `Board::create()` set corners for marker in `objPoints` with "CCW" order:
![image](https://user-images.githubusercontent.com/22337800/154343015-3c2e7d0b-271c-41e0-b768-d0a421e5b678.png)
But `detectMarkers()` find markers and reordered corners to "CW" order:
![image](https://user-images.githubusercontent.com/22337800/154343485-e2cba76d-c4a3-4c68-9628-b728a94f793f.png)

This makes coordinate systems not consistent and makes the code hard to read

alalek commented on modules/aruco/test/test_aruco_utils.hpp:
```diff
@@ -0,0 +1,101 @@
+namespace opencv_test {
+    namespace {
+
+        static inline double deg2rad(double deg) { return deg * CV_PI / 180.; }
```
avoid indentation in namespaces

AleksandrPanov replied:
fixed

AleksandrPanov replied:
@alalek thanks for the review)
do you have any more questions?

AleksandrPanov commented:
@alalek thanks for the review)
do you have any more questions?

asmorkalov commented on modules/aruco/src/aruco.cpp:
```diff
@@ -1782,9 +1787,18 @@ void drawDetectedMarkers(InputOutputArray _image, InputArrayOfArrays _corners,
+    axis.push_back(Point3f(0.f, 0.f, 0.f));
+    axis.push_back(Point3f(length, 0.f, 0.f));
+    axis.push_back(Point3f(0.f, length, 0.f));
+    axis.push_back(Point3f(0.f, 0.f, -length));
```
why `-length`? Your function differs from standard drawFrameAxes with opposite direction of the third axis.

AleksandrPanov replied:
This was added to improve appearance. The drawAxis() function doesn't make sense, as it is just calls drawFrameAxes().

asmorkalov commented on modules/aruco/test/test_aruco_utils.hpp:
```diff
@@ -0,0 +1,128 @@
+    axis.push_back(Point3f(offset, offset, 0.f));
+    axis.push_back(Point3f(length+offset, offset, 0.f));
+    axis.push_back(Point3f(offset, length+offset, 0.f));
+    axis.push_back(Point3f(offset, offset, -length));
```
The same, the result axis direction is opposite to what standard drawFrameAxes does.

AleksandrPanov replied:
fixed

AleksandrPanov commented:
I checked the coordinate system axis from pose estimation:

Chessboard pose estimation:
![image](https://user-images.githubusercontent.com/22337800/156664964-a163045e-8000-436a-90ba-22added9001b.png)

ArUco board pose estimation (after fix):
![image](https://user-images.githubusercontent.com/22337800/156665473-6347323f-9d59-4adc-b582-e54d41f15d08.png)

ChArUco board pose estimation (after fix):
![image](https://user-images.githubusercontent.com/22337800/156665634-d6aab681-7039-49ce-b08e-47e8ffbf9776.png)

The `drawAxis()` function doesn't make sense, as it is just calls `drawFrameAxes()` and because `drawAxis()` was removed.

Also `drawCustomAxis()` was added:
![image](https://user-images.githubusercontent.com/22337800/156666308-cc482112-2c47-4b82-8b2b-15e6eab41ea1.png)

[poseEstimate.py.txt](https://github.com/opencv/opencv_contrib/files/8181631/poseEstimate.py.txt)
Use this image - aruco\tutorials\aruco_calibration\images\img_00.jpg:
![img_00](https://user-images.githubusercontent.com/22337800/156666547-7d9fc252-532d-4a1a-97b5-5d2bd834080f.jpg)

alalek commented:
/cc @catree It would be great if you could take a look on this too

catree commented:
First very quick look.

---

Not sure to understand the purpose of `cv::aruco::drawCustomAxis()` vs `cv::drawFrameAxes()`:
- `void drawCustomAxis(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3, Point3f center = Point3f(0.f, 0.f, 0.f), Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f))`
- `void drawFrameAxes(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3)`

About the shifted center point `Point3f center = Point3f(0.f, 0.f, 0.f)`, I think what you are doing is:

![image](https://user-images.githubusercontent.com/8229425/156840632-a05fe98b-1e0d-4d82-a590-31ceae0d2e54.png)

But this is equivalent to:

![image](https://user-images.githubusercontent.com/8229425/156840755-d78bcc8f-182b-46ae-a3ab-03f0aead686c.png)

About the `Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f)`, is there really a need to use different sizes for the frame axes?

```python
cv.aruco.drawCustomAxis(img, K, D, rvecs, tvecs, 1., 3, (0, 7, 0), (1, -1, -1))
```

Should be equivalent to:

![image](https://user-images.githubusercontent.com/8229425/156841779-7d9e6cfc-55e8-4101-bae3-9e024e23876d.png)

![image](https://user-images.githubusercontent.com/8229425/156842249-f220cbf1-ae04-477e-a0a1-2ab1e4ad42dc.png)

AleksandrPanov commented:
> First very quick look.
> 
> Not sure to understand the purpose of `cv::aruco::drawCustomAxis()` vs `cv::drawFrameAxes()`:
> 
> * `void drawCustomAxis(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3, Point3f center = Point3f(0.f, 0.f, 0.f), Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f))`
> * `void drawFrameAxes(InputOutputArray image, InputArray cameraMatrix, InputArray distCoeffs, InputArray rvec, InputArray tvec, float length, int thickness=3)`
> 
> About the shifted center point `Point3f center = Point3f(0.f, 0.f, 0.f)`, I think what you are doing is:
> 
> ![image](https://user-images.githubusercontent.com/8229425/156840632-a05fe98b-1e0d-4d82-a590-31ceae0d2e54.png)
> 
> But this is equivalent to:
> 
> ![image](https://user-images.githubusercontent.com/8229425/156851184-2741296e-63ab-492e-85e5-3728ac2512a1.png)
> 
> About the `Point3f axisCoefficient = Point3f(1.f, 1.f, 1.f)`, is there really a need to use different sizes for the frame axes?
> 
> ```python
> cv.aruco.drawCustomAxis(img, K, D, rvecs, tvecs, 1., 3, (0, 7, 0), (1, -1, -1))
> ```
> 
> Should be equivalent to:
> 
> ![image](https://user-images.githubusercontent.com/8229425/156841779-7d9e6cfc-55e8-4101-bae3-9e024e23876d.png)
> 
> ![image](https://user-images.githubusercontent.com/8229425/156842249-f220cbf1-ae04-477e-a0a1-2ab1e4ad42dc.png)

@catree the direction/position of the axes changed after objPoints fix. A `drawCustomAxis()` has been added to draw the axes in the old style to avoid reworking the tutorial. Do I understand correctly that you are proposing to use transformations with rvecs, tvecs vectors to get the desired axis direction/position?

catree commented:
I am not fan of the `drawCustomAxis()` because it feels unnecessary for me. I would prefer some good API for handling SE3/SO3 transformations with some good tutorials. You can also shoot yourself in the foot if you use `axisCoefficient=(-1,0,0)` or any other non valid values.

What about [`composeRT()`](https://docs.opencv.org/4.5.5/d9/d0c/group__calib3d.html#ga044b03d81f99370db05c52c3d0b46599), can it do the same job?

AleksandrPanov commented:
> I am not fan of the `drawCustomAxis()` because it feels unnecessary for me. I would prefer some good API for handling SE3/SO3 transformations with some good tutorials. You can also shoot yourself in the foot if you use `axisCoefficient=(-1,0,0)` or any other non valid values.
> 
> Moreover, the extra functionality of `drawCustomAxis()` is not used in your PR (only for debugging purposes), no?
> 
> What about [`composeRT()`](https://docs.opencv.org/4.5.5/d9/d0c/group__calib3d.html#ga044b03d81f99370db05c52c3d0b46599), can it do the same job?

@catree, thx)
I removed `drawCustomAxis()` and added `cv::drawFrameAxes()` link to docs and replaced old images with new ones. The tutorial now shows the correct direction and position of the axes.

AleksandrPanov commented:
@catree, I think that good API for handling SE3/SO3 transformations should add into another PR. This PR fixes only ArUco problems.

alalek reviewed (approved):
Thank you :+1:

mabl commented:
It looks like this change makes `interpolateCornersCharuco`  return a different number for board corners. [See image here](https://github.com/opencv/opencv_contrib/commit/56d492cfecb7aadaca79a9494a33ad1c12798e5e#diff-de110295d0b9a89eb14a5aedc00a241012295be05e4ab60776f3ec94f5227058). Before this, this corner id mapped to the index of `board->chessboardCorners`. 

Is this change intentional? If so, how can I recover the corresponding index for `board->chessboardCorners` from [interpolateCornersCharuco](https://docs.opencv.org/3.4/d9/d6a/group__aruco.html#gadcc5dc30c9ad33dcf839e84e8638dcd1) markerIds, i.e. the list of identifiers for each marker in corners.
